src/sensor_ads1220.c

```diff
@@ -0,0 +1,206 @@
+ * ADS1220 Sensor Support
+ ****************************************************************/
+#define MAX_SPI_READ_TIME timer_from_us(150)
+
+static inline uint8_t
+is_flag_set(const uint8_t mask, struct ads1220_adc *ads1220)
+{
+    return !!(mask & ads1220->flags);
+}
+
+static inline void
+set_flag(uint8_t mask, struct ads1220_adc *ads1220)
+{
+    ads1220->flags |= mask;
+}
+
+static inline void
+clear_flag(uint8_t mask, struct ads1220_adc *ads1220)
+{
+    ads1220->flags &= ~mask;
+}
+
+// Event handler that wakes wake_ads1220() periodically
+static uint_fast8_t
+ads1220_event(struct timer *timer)
+{
+    struct ads1220_adc *ads1220 = container_of(timer, struct ads1220_adc,
+                                                timer);
+    set_flag(FLAG_PENDING, ads1220);
+    sched_wake_task(&wake_ads1220);
+    return SF_DONE;
+}
+
+// Helper code to reschedule the ads1220_event() timer
+static void
+ads1220_reschedule_timer(struct ads1220_adc *ads1220)
+{
+    irq_disable();
+    set_flag(FLAG_PENDING, ads1220);
+    ads1220->timer.waketime = timer_read_time() + ads1220->rest_ticks;
+    sched_add_timer(&ads1220->timer);
+    irq_enable();
+}
+
+int8_t
+ads1220_is_data_ready(struct ads1220_adc *ads1220) {
+    return gpio_in_read(ads1220->data_ready) == 0;
+}
+
+// Add a measurement to the buffer
+static void
+add_sample(struct ads1220_adc *ads1220, uint_fast32_t counts)
+{
+    ads1220->sb.data[ads1220->sb.data_count] = counts;
+    ads1220->sb.data[ads1220->sb.data_count + 1] = counts >> 8;
+    ads1220->sb.data[ads1220->sb.data_count + 2] = counts >> 16;
+    ads1220->sb.data[ads1220->sb.data_count + 3] = counts >> 24;
+    ads1220->sb.data_count += BYTES_PER_SAMPLE;
+}
+
+static void
+flush_samples(struct ads1220_adc *ads1220, uint8_t oid)
+{
+    if ((ads1220->sb.data_count + BYTES_PER_SAMPLE) >
+            ARRAY_SIZE(ads1220->sb.data)) {
+        sensor_bulk_report(&ads1220->sb, oid);
+    }
+}
+
+// ADS1220 ADC query
+void
+ads1220_read_adc(struct ads1220_adc *ads1220, uint8_t oid)
+{
+    if (!ads1220_is_data_ready(ads1220)) {
+        ads1220_reschedule_timer(ads1220);
+        return;
+    }
+
+    // data is ready
+    uint8_t msg[3] = {0, 0, 0};
+    uint32_t start_time = timer_read_time();
+    spidev_transfer(ads1220->spi, 1, sizeof(msg), msg);
+    uint32_t time_diff = timer_read_time() - start_time;
+
+    if (time_diff >= MAX_SPI_READ_TIME) {
+        // some IRQ delayed this read so much that its unusable
+        shutdown("ADS1220 read timing error, read took too long");
+    }
+
+    // extend 2's complement 24 bits to 32 bits
+    int32_t counts_original = (msg[0] << 16) | (msg[1] << 8) | msg[2];
+    uint32_t m = 1u << (24 - 1);
+    int32_t counts = (counts_original ^ m) - m;
+
+    if (counts == -1) {
+        shutdown("ADS1220: Possible bad read");
+    }
+
+    if (counts >= 0x800000) {
+        counts |= 0xFF000000;
+        shutdown("ADS1220: Invalid Counts");
+    }
+    add_sample(ads1220, counts);
```

Comment:
Using 4 bytes for a 3-byte reading seems wasteful of cpu and bandwidth resources. It would seem better to encode this into the bulk_sensor report using 3-bytes.

Separately, a sample read error could be encoded as 0x800000, and if the sensor actually reports 0x800000 we can turn that into 0x800001 to avoid a conflict.

Reply:
Agree we can encode errors as "-0", 0x800000 or 0x80000000.

To support the endstop on the MCU the sample has to be converted to 32 bits. I didn't want to do this conversion and not see the result in Python. I was/am worried that the conversion is done correctly. Sending back the 32 bits gives me that visibility.

Neither of these sensors can sample so fast that I was worried about the overhead. On the ADS1263, it can sample at 32Khz, but it is 32 bits anyway. The ADXL345 is sending back more data per measurement that this and sampling faster. So I judged the overhead worth the visibility.

Reply:
FYI, adxl345 has 3200 samples per second, and 5 bytes per sample are sent to the host.  (So, 10x slower max sampling rate than ADS1283).

Reply:
the max sampling rate on this chip, ADS1220 is 2000 samples per second. So less than ADXL345.